### PR TITLE
Chicken: fix env vars

### PR DESCRIPTION
**Pull request by zilti**
## Description of changes

Adjustment for Chicken's environment variables to properly allow for overrides for manual egg installations.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [X] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by zilti** [comment]
Maintainer pings: @corngood @nagy @konst-aa

**Comment by corngood** [comment]
Can you explain a bit more about the use case that's broken?

**Comment by zilti** [comment]
More than one thing is broken really, but the main thing is: when you want to use `chicken-install`, you have to set `CHICKEN_REPOSITORY_PATH` and `CHICKEN_INSTALL_REPOSITORY`. Doing the former currently overrides the repository path completely, which means Chicken won't find its internal libraries anymore. So manually installing an egg and trying to use it will currently break Chicken. This patch ensures that Chicken will always also have its original `CHICKEN_REPOSITORY_PATH`, making it work again.

**Comment by zilti** [comment]
To put it into a code example, this would fail currently, but is fixed with the patch:

```
export CHICKEN_INSTALL_REPOSITORY=$HOME/.local/chicken/eggs
export CHICKEN_REPOSITORY_PATH=$HOME/.local/chicken/eggs
chicken-install utf8
```

**Review by corngood — COMMENTED** [review]
A couple of general things:

- merge commit should be removed
- commit message should contain a bit more detail about what's changing

In theory this is a breaking change because it will stop someone from fully overriding `CHICKEN_REPOSITORY_PATH`.  However, as long as you can prefix to it, that's probably not a serious concern.

File: pkgs/development/compilers/chicken/5/chicken.nix [hunk 1]
```diff
@@ -66,6 +66,16 @@ stdenv.mkDerivation (finalAttrs: {
     command = "csi -version";
   };
 
+  postInstall = ''
+    for f in $out/bin/*
+    do
+      wrapProgram $f \
+        --prefix CHICKEN_REPOSITORY_PATH : "$out/lib/chicken/${toString finalAttrs.binaryVersion}:$CHICKEN_REPOSITORY_PATH" \
```
[the remark this answers is not in the document]
Comment (zilti) [reply pkgs/development/compilers/chicken/5/chicken.nix:73]: No, it adds the user-set `$CHICKEN_REPOSITORY_PATH` to Chicken's default path. You can see the same approach in eggDerivation.nix.

File: pkgs/development/compilers/chicken/5/chicken.nix [hunk 2]
```diff
@@ -66,6 +66,16 @@ stdenv.mkDerivation (finalAttrs: {
...
+        --prefix CHICKEN_INSTALL_PREFIX : "$out" \
```
[the remark this answers is not in the document]
Comment (zilti) [reply pkgs/development/compilers/chicken/5/chicken.nix:74]: I explained above what happens when it is not set. And no, it is not meant to be a single path, but a collection of search paths, like e.g. `$PATH`. Again, the same approach is already used in eggDerivation.nix.

[the remark this answers is not in the document]
Comment (corngood) [reply pkgs/development/compilers/chicken/5/chicken.nix:74]: I was referring to `CHICKEN_INSTALL_PREFIX`.  I'm pretty sure it's meant to be a single path.

**Comment by corngood** [comment]
> Doing the former currently overrides the repository path completely, which means Chicken won't find its internal libraries anymore.

Can you explain how to repro this?

`CHICKEN_REPOSITORY_PATH=/var/empty $(nix-build -A chicken --no-out-link)/bin/csi -e '(import (chicken base))'` seems to work.

**Comment by zilti** [comment]
Well... this is just all an utter mess.

**Comment by zilti** [comment]
Seriously. All I wanted was to be able to evaluate Chicken code using Geiser, and it just deteriorated into this massive yakshaving fest. I don't have the nerves anymore. May someone else fix this.

**Comment by nagy** [comment]
> Well... this is just all an utter mess.

This is just how things go normally.  You have to explain elaborately what is exactly the case that you want to fix.  Your title "fix env vars" is very generic.  I also have a bit trouble understanding your case.  If you are talking about manually adding eggs, that are not packaged in nixpkgs yet, then you may have a point.  All upstream eggs are already packaged and ready to use.

I am sure we could gladly use your help but have to show what is currently not working.  The best scenario would even be to provide a testcase in form of code, not only would that help us to understand your situation but it could help us see if we break such functionality in the future.


> To put it into a code example, this would fail currently, but is fixed with the patch:
>  export CHICKEN_INSTALL_REPOSITORY=$HOME/.local/chicken/eggs
>  export CHICKEN_REPOSITORY_PATH=$HOME/.local/chicken/eggs
>  chicken-install utf8


You should not need to imperatively install eggs, which are already packaged. For example

```
nix-shell -p chicken chickenPackages.chickenEggs.utf8 --run "csi -e '(import utf8)'"
```

works for me and with that csi I am able to access the utf8 package.
For manual packages, you may have a point but anything that just involves already packaged eggs and evaluating code in geiser on them already works.

**Comment by nagy** [comment]
> There's a legit problem here, 

I agree, we should support the case of manually calling `chicken-install` most importantly for private eggs or during the development. But we should discourage the installing of already packaged eggs via this method because if any of the dependencies contain links to native libraries, the whole thing will not work.

> prefix vs. suffix?

If we ever decide to continue with this, I would prefer suffix. That way, only missing dependencies get filled in while the user is able to place a modified version of an already existing egg in that directory and see the change directly.

> Should we make these defaults and attempt to make chicken-install work out of the box?

We should follow upstream as close as possible on path choices.

https://wiki.call-cc.org/man/5/Extension%20tools#changing-the-repository-location

here it says `/usr/local/lib/chicken/9/` could be a choice so that would translate to `$HOME/.local/chicken/${toString chicken.binaryVersion}/"` ? Maybe it should even respect `XDG_DATA_DIRS`.

**Comment by zilti** [comment]
Yeah the problems here run deeper than just that env var. I'm honestly a bit confused. 

About Chicken looking at `XDG_DATA_DIRS` as an addition to `CHICKEN_REPOSITORY_PATH`, I will bring that up. It might make it into Chicken 6; I doubt there'll be all that many additions to Chicken 5 in the future.

> Should we make these defaults and attempt to make chicken-install work out of the box? The default paths will never be writeable in nix.

That is a tricky one. I suppose these could be set to e.g. `${XDG_DATA_HOME}/chicken/${toString chicken.binaryVersion}`. But the issue on top of that one is that eggs that install binaries install them by default to `${CHICKEN_INSTALL_PREFIX}/bin`.

This should be changed upstream; I do not know what the canonical directory for this is, though. I assumed it is `$HOME/.local/bin`, and indeed this seems to be the recommendation of the XDG spec. This would require a new upstream env variable, maybe `CHICKEN_INSTALL_BINARY_REPOSITORY`.

**Pull request closed without merging**